// @foxpage/foxpage-shared/lib/common/Messages.js
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatMessages = exports.stringifyMessages = exports.Messages = void 0;
const isError_1 = __importDefault(require("lodash/isError"));
const utils_1 = require("./utils");
class Messages extends Array {
    get hasError() {
        return this.some(isError_1.default);
    }
    formate(includeStack) {
        return formatMessages(this, includeStack);
    }
    stringify(splitter = ' | ') {
        return stringifyMessages(this, splitter);
    }
    toJSON() {
        return this.formate();
    }
}
exports.Messages = Messages;
function stringifyMessages(messages = [], splitter = ' | ') {
    return formatMessages(messages).join(splitter);
}
exports.stringifyMessages = stringifyMessages;
function formatMessages(messages = [], includeStack = false) {
    return messages.map(msg => (isError_1.default(msg) ? (includeStack ? msg.stack : msg.message) : msg)).filter(utils_1.isNotNill);
}
exports.formatMessages = formatMessages;
