// @foxpage/foxpage-core/lib/structure/creators.js
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createStructureWithFactory = exports.createStructureComponent = exports.createStructureDSL = void 0;
const shortid_1 = __importDefault(require("shortid"));
function createStructureDSL(name, props = {}, partial = {}) {
    const structure = Object.assign({ id: `create_${name}_${shortid_1.default()}`, props, type: 'react.component', name, label: '', version: '' }, partial);
    return structure;
}
exports.createStructureDSL = createStructureDSL;
function createStructureComponent(structure) {
    const component = {
        name: structure.name,
        browserURL: '',
        debugURL: '',
        nodeURL: '',
        cssURL: '',
        supportSSR: true,
        meta: {},
    };
    return component;
}
exports.createStructureComponent = createStructureComponent;
function createStructureWithFactory(factory, props = {}, partial = {}) {
    const label = typeof factory === 'string' ? factory : factory.displayName || 'Unknown';
    const group = typeof factory === 'string' ? 'tag' : 'builtin';
    const name = `${String(group)}.${label}`;
    const structure = createStructureDSL(name, {}, partial);
    structure.props = props;
    const component = createStructureComponent(structure);
    component.factory = factory;
    return { structure, component };
}
exports.createStructureWithFactory = createStructureWithFactory;
