// tsoa-custom-decorators/dist/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var example_1 = require("./decorators/example");
exports.Example = example_1.Example;
var parameter_1 = require("./decorators/parameter");
exports.Request = parameter_1.Request;
exports.Query = parameter_1.Query;
exports.Path = parameter_1.Path;
exports.Body = parameter_1.Body;
exports.BodyProp = parameter_1.BodyProp;
exports.Header = parameter_1.Header;
var methods_1 = require("./decorators/methods");
exports.Post = methods_1.Post;
exports.Get = methods_1.Get;
exports.Patch = methods_1.Patch;
exports.Delete = methods_1.Delete;
exports.Put = methods_1.Put;
var tags_1 = require("./decorators/tags");
exports.Tags = tags_1.Tags;
var route_1 = require("./decorators/route");
exports.Route = route_1.Route;
var security_1 = require("./decorators/security");
exports.Security = security_1.Security;
var controller_1 = require("./interfaces/controller");
exports.Controller = controller_1.Controller;
var response_1 = require("./decorators/response");
exports.Response = response_1.Response;
exports.SuccessResponse = response_1.SuccessResponse;
var templateHelpers_1 = require("./routeGeneration/templateHelpers");
exports.ValidateParam = templateHelpers_1.ValidateParam;
__export(require("./decorators/validations"));
