// @foxpage/foxpage-plugin-unpkg/lib/functions.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
exports.formatResponse = exports.getPackageFoxpage = exports.getPackageManifest = exports.searchNpmPackage = void 0;
var axios_1 = require("axios");
var _ = require("lodash");
var ProxyAgent = require("proxy-agent");
var UNPKG_HOST = 'https://www.unpkg.com/';
/**
 * Search from npm by the special text, only response 20 packages
 * @param  {{name?:string}} options
 * @returns Promise
 */
var searchNpmPackage = function (options) { return __awaiter(void 0, void 0, void 0, function () {
    var proxyConfig, proxyAgent, data;
    var _a, _b;
    return __generator(this, function (_c) {
        switch (_c.label) {
            case 0:
                if (!options.packageName || !((_a = options.groupConfig) === null || _a === void 0 ? void 0 : _a.resourceScope)) {
                    return [2 /*return*/, []];
                }
                proxyConfig = {};
                if (options.proxy) {
                    proxyAgent = new ProxyAgent(options.proxy);
                    proxyConfig = {
                        httpAgent: proxyAgent,
                        httpsAgent: proxyAgent
                    };
                }
                return [4 /*yield*/, (0, axios_1["default"])(Object.assign({
                        method: 'get',
                        url: 'https://registry.npmjs.com/-/v1/search?text=' +
                            _.trim((((_b = options.groupConfig) === null || _b === void 0 ? void 0 : _b.resourceScope) || options.packageName))
                            + '&size=500'
                    }, proxyConfig))];
            case 1:
                data = (_c.sent()).data;
                return [2 /*return*/, _.map(data.objects || [], function (pkg) { return _.pick(pkg.package || {}, ['name', 'scope', 'version']); })];
        }
    });
}); };
exports.searchNpmPackage = searchNpmPackage;
/**
 * Get package manifest file path from unpkg
 * @param  {{name:string;version:string;}} options
 * @returns Promise
 */
var getPackageManifest = function (options) { return __awaiter(void 0, void 0, void 0, function () {
    var data, err_1;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                _a.trys.push([0, 2, , 3]);
                return [4 /*yield*/, (0, axios_1["default"])({
                        method: 'get',
                        url: UNPKG_HOST + options.name + '@' + options.version + '/dist/manifest.json'
                    })];
            case 1:
                data = (_a.sent()).data;
                return [2 /*return*/, Object.assign({ manifest: data }, options)];
            case 2:
                err_1 = _a.sent();
                return [2 /*return*/, Object.assign({ manifest: {} }, options)];
            case 3: return [2 /*return*/];
        }
    });
}); };
exports.getPackageManifest = getPackageManifest;
/**
 * Get package foxpage json file data
 * @param  {{name:string;version:string}} options
 */
var getPackageFoxpage = function (options) { return __awaiter(void 0, void 0, void 0, function () {
    var data, err_2;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                _a.trys.push([0, 2, , 3]);
                return [4 /*yield*/, (0, axios_1["default"])({
                        method: 'get',
                        url: UNPKG_HOST + options.name + '@' + options.version + '/dist/foxpage.json'
                    })];
            case 1:
                data = (_a.sent()).data;
                return [2 /*return*/, Object.assign({ foxpage: (data === null || data === void 0 ? void 0 : data.foxpage) || {} }, options)];
            case 2:
                err_2 = _a.sent();
                return [2 /*return*/, Object.assign({ foxpage: {} }, options)];
            case 3: return [2 /*return*/];
        }
    });
}); };
exports.getPackageFoxpage = getPackageFoxpage;
/**
 * format response resource structure to {name,version,dirName,files}
 * @param  {UnpkgManifest[]} manifestList
 * @returns IndexContentPkg
 */
var formatResponse = function (manifestList, pkgFoxpageData) {
    var resources = [];
    var pkgFoxpageObject = _.keyBy(pkgFoxpageData, function (pkg) {
        return pkg.name + '@' + pkg.version;
    });
    manifestList.forEach(function (pkg) {
        var _a;
        if (!_.isEmpty(pkg.manifest)) {
            var manifestFile = {};
            for (var key in pkg.manifest) {
                var tmpFile = transformArrToObject(key.split('/'), _.startsWith(pkg.manifest[key], '/') ? '' : '/' + pkg.manifest[key], pkg);
                manifestFile = _.merge(manifestFile, tmpFile);
            }
            resources.push({
                name: pkg.name,
                version: pkg.version,
                foxpage: Object.assign({ resourceName: _.last(pkg.name.split('/')) }, (_a = pkgFoxpageObject[pkg.name + '@' + pkg.version]) === null || _a === void 0 ? void 0 : _a.foxpage),
                files: manifestFile
            });
        }
    });
    return resources;
};
exports.formatResponse = formatResponse;
/**
 * transform array to recursive object, eg. ['a','b','c'] => {a: {b:c}}
 * @param  {string[]} keyArr
 * @param  {string} value
 * @returns any
 */
var transformArrToObject = function (keyArr, value, pkg) {
    var _a;
    var itemObject = {};
    for (var _i = 0, _b = _.reverse(keyArr); _i < _b.length; _i++) {
        var key = _b[_i];
        if (key === _.first(keyArr)) {
            itemObject[key] = '/' + pkg.name + '@' + pkg.version + '/dist' + value;
        }
        else {
            itemObject = (_a = {}, _a[key] = itemObject, _a);
        }
    }
    return itemObject;
};
