// @foxpage/foxpage-plugin-unpkg/lib/main.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
var _ = require("lodash");
var functions_1 = require("./functions");
var PLUGIN_NAME = 'unpkg';
var Unpkg = function () {
    return {
        name: 'foxpage-unpkg',
        visitor: {
            resourceList: function (options) { return __awaiter(void 0, void 0, void 0, function () {
                var npmPkgList, pkgManifestList, pkgFoxpageData, _i, _a, pkgBlock, blockPkgManifestList, blockPkgFoxpageData, groupResources;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            if (!options.type || options.type.toLowerCase() !== PLUGIN_NAME) {
                                return [2 /*return*/, { type: PLUGIN_NAME, group: '', packages: [] }];
                            }
                            return [4 /*yield*/, (0, functions_1.searchNpmPackage)(options)];
                        case 1:
                            npmPkgList = _b.sent();
                            pkgManifestList = [];
                            pkgFoxpageData = [];
                            if (!(npmPkgList.length > 0)) return [3 /*break*/, 6];
                            _i = 0, _a = _.chunk(npmPkgList, 20);
                            _b.label = 2;
                        case 2:
                            if (!(_i < _a.length)) return [3 /*break*/, 6];
                            pkgBlock = _a[_i];
                            return [4 /*yield*/, Promise.all(pkgBlock.map(function (pkg) { return (0, functions_1.getPackageManifest)(pkg); }))];
                        case 3:
                            blockPkgManifestList = _b.sent();
                            return [4 /*yield*/, Promise.all(pkgBlock.map(function (pkg) { return (0, functions_1.getPackageFoxpage)(pkg); }))];
                        case 4:
                            blockPkgFoxpageData = _b.sent();
                            pkgManifestList = pkgManifestList.concat(blockPkgManifestList);
                            pkgFoxpageData = pkgFoxpageData.concat(blockPkgFoxpageData);
                            _b.label = 5;
                        case 5:
                            _i++;
                            return [3 /*break*/, 2];
                        case 6:
                            groupResources = { type: PLUGIN_NAME, group: '', packages: [] };
                            if (pkgManifestList.length > 0) {
                                groupResources.packages = (0, functions_1.formatResponse)(pkgManifestList, pkgFoxpageData);
                            }
                            return [2 /*return*/, groupResources];
                    }
                });
            }); },
            resourceRemoteUrl: function (options) {
                var _a;
                return {
                    type: PLUGIN_NAME,
                    url: 'https://registry.npmjs.com/-/v1/search?text=' +
                        _.trim((((_a = options.groupConfig) === null || _a === void 0 ? void 0 : _a.resourceScope) || options.packageName))
                };
            }
        },
        options: {}
    };
};
exports["default"] = Unpkg;
