// @foxpage/foxpage-plugin-function-parse/lib/index.js
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * register function parser plugin
 * @returns function parser
 */
const functionParser = ({ api }) => {
    return {
        visitor: {
            registerVariableParser: () => {
                return {
                    type: 'data.function.call',
                    parse(variable, context) {
                        const { props: { function: code, args = [] }, } = variable;
                        if (code) {
                            const fn = (api === null || api === void 0 ? void 0 : api.evalWithScope) ? api.evalWithScope({ $this: context }, code)
                                : '';
                            if (typeof fn === 'function') {
                                const result = fn.call(context, ...args);
                                return result;
                            }
                        }
                        throw new Error(`${variable.name} function is correct`);
                    },
                };
            },
        },
    };
};
exports.default = functionParser;
