// @foxpage/foxpage-core/lib/merger/strategy/type.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.isCombine = exports.isReplace = exports.isMatchByExtend = exports.isMatchByName = exports.isMatchById = exports.COMBINE_STRATEGY = exports.REPLACE_STRATEGY = exports.MATCH_BY_EXTEND = exports.MATCH_BY_NAME = exports.MATCH_BY_ID = exports.MergeStrategy = void 0;
/**
 * merge strategy
 */
var MergeStrategy;
(function (MergeStrategy) {
    /**
     * replace
     * default equals REPLACE_BY_ID
     */
    MergeStrategy[MergeStrategy["REPLACE"] = 1] = "REPLACE";
    /**
     * matched by id then replace
     */
    MergeStrategy[MergeStrategy["REPLACE_BY_ID"] = 2] = "REPLACE_BY_ID";
    /**
     * matched by name then replace
     */
    MergeStrategy[MergeStrategy["REPLACE_BY_NAME"] = 3] = "REPLACE_BY_NAME";
    /**
     * combine
     * default equals COMBINE_BY_ID
     */
    MergeStrategy[MergeStrategy["COMBINE"] = 10] = "COMBINE";
    /**
     * matched by id then combine
     */
    MergeStrategy[MergeStrategy["COMBINE_BY_ID"] = 11] = "COMBINE_BY_ID";
    /**
     * matched by name then combine
     */
    MergeStrategy[MergeStrategy["COMBINE_BY_NAME"] = 12] = "COMBINE_BY_NAME";
    /**
     * matched by extend(inherit) the combine
     */
    MergeStrategy[MergeStrategy["COMBINE_BY_EXTEND"] = 13] = "COMBINE_BY_EXTEND";
    // /**
    //  * matched by position then combine
    //  */
    // COMBINE_BY_POSITION = 14,
})(MergeStrategy = exports.MergeStrategy || (exports.MergeStrategy = {}));
// match
exports.MATCH_BY_ID = [
    MergeStrategy.COMBINE,
    MergeStrategy.COMBINE_BY_ID,
    MergeStrategy.REPLACE,
    MergeStrategy.REPLACE_BY_ID,
];
exports.MATCH_BY_NAME = [MergeStrategy.COMBINE_BY_NAME, MergeStrategy.REPLACE_BY_NAME];
exports.MATCH_BY_EXTEND = [MergeStrategy.COMBINE_BY_EXTEND];
// merge
exports.REPLACE_STRATEGY = [MergeStrategy.REPLACE, MergeStrategy.REPLACE_BY_ID, MergeStrategy.REPLACE_BY_NAME];
exports.COMBINE_STRATEGY = [
    MergeStrategy.COMBINE,
    MergeStrategy.COMBINE_BY_ID,
    MergeStrategy.COMBINE_BY_NAME,
    MergeStrategy.COMBINE_BY_EXTEND,
];
// match
exports.isMatchById = (strategy) => {
    return exports.MATCH_BY_ID.indexOf(strategy) > -1;
};
exports.isMatchByName = (strategy) => {
    return exports.MATCH_BY_NAME.indexOf(strategy) > -1;
};
exports.isMatchByExtend = (strategy) => {
    return exports.MATCH_BY_EXTEND.indexOf(strategy) > -1;
};
// merge
exports.isReplace = (strategy) => {
    return exports.REPLACE_STRATEGY.indexOf(strategy) > -1;
};
exports.isCombine = (strategy) => {
    return exports.COMBINE_STRATEGY.indexOf(strategy) > -1;
};
