// tsoa-custom-decorators/dist/metadataGeneration/controllerGenerator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ts = require("typescript");
var methodGenerator_1 = require("./methodGenerator");
var ControllerGenerator = (function () {
    function ControllerGenerator(node, decoratorsSchema /* = ['Route', 'Controller', 'JsonController']*/) {
        this.node = node;
        this.decoratorsSchema = decoratorsSchema; /* = ['Route', 'Controller', 'JsonController']*/
        this.pathValue = this.getControllerRouteValue(node);
    }
    ControllerGenerator.prototype.IsValid = function () {
        return !!this.pathValue || this.pathValue === '';
    };
    ControllerGenerator.prototype.Generate = function () {
        if (!this.node.parent) {
            throw new Error('Controller node doesn\'t have a valid parent source file.');
        }
        if (!this.node.name) {
            throw new Error('Controller node doesn\'t have a valid name.');
        }
        var sourceFile = this.node.parent.getSourceFile();
        return {
            location: sourceFile.fileName,
            methods: this.buildMethods(),
            name: this.node.name.text,
            path: this.pathValue || ''
        };
    };
    ControllerGenerator.prototype.buildMethods = function () {
        var _this = this;
        if (!this.node.name) {
            throw new Error('Controller node doesn\'t have a valid name.');
        }
        return this.node.members
            .filter(function (m) { return m.kind === ts.SyntaxKind.MethodDeclaration; })
            .map(function (m) { return new methodGenerator_1.MethodGenerator(m, _this.decoratorsSchema); })
            .filter(function (generator) { return generator.IsValid(); })
            .map(function (generator) { return generator.Generate(); });
    };
    ControllerGenerator.prototype.getControllerRouteValue = function (node) {
        return this.getControllerDecoratorValue(node, this.decoratorsSchema.controllersDecorators.map(function (d) { return d.name; }), '');
    };
    ControllerGenerator.prototype.getControllerDecoratorValue = function (node, decoratorName, defaultValue) {
        if (!node.decorators) {
            return undefined;
        }
        var matchedAttributes = node.decorators
            .map(function (d) { return d.expression; })
            .filter(function (expression) {
            var subExpression = expression.expression;
            var result = subExpression.text === decoratorName;
            if (decoratorName instanceof Array) {
                result = !!decoratorName.find(function (decoratorName) { return decoratorName === subExpression.text; });
            }
            return result;
        });
        if (!matchedAttributes.length) {
            return undefined;
        }
        if (matchedAttributes.length > 1) {
            throw new Error("A controller can only have a single 'decoratorName' decorator in `" + this.node.name.text + "` class.");
        }
        var value = matchedAttributes[0].arguments[0];
        return value ? value.text : defaultValue;
    };
    return ControllerGenerator;
}());
exports.ControllerGenerator = ControllerGenerator;
