// tsoa-custom-decorators/dist/metadataGeneration/defaultDecoratorsSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_DECORATORS_SCHEMA = {
    controllersDecorators: [
        {
            name: "Route",
            tsoaDecorator: "Route"
        },
        {
            name: "Controller",
            tsoaDecorator: "Route"
        },
        {
            name: "JsonController",
            tsoaDecorator: "Route"
        }
    ],
    methodDecorators: [
        {
            name: "get",
            tsoaDecorator: "get"
        },
        {
            name: "post",
            tsoaDecorator: "post"
        },
        {
            name: "patch",
            tsoaDecorator: "patch"
        },
        {
            name: "delete",
            tsoaDecorator: "delete"
        },
        {
            name: "put",
            tsoaDecorator: "put"
        }
    ],
    parameterDecorators: [
        {
            name: "Request",
            tsoaDecorator: "Request"
        },
        {
            name: "Body",
            tsoaDecorator: "Body"
        },
        {
            name: "BodyProp",
            tsoaDecorator: "BodyProp"
        },
        {
            name: "Header",
            tsoaDecorator: "Header"
        },
        {
            name: "Query",
            tsoaDecorator: "Query"
        },
        {
            name: "Path",
            tsoaDecorator: "Path"
        }
    ]
};
